File: baekjoon/src/tags/stack/No9012.java
package tags.stack;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

/**
 * https://www.acmicpc.net/problem/9012
 */
public class No9012 {
    public static void command(String command) {
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < command.length(); i++)
            stack.push(command.charAt(i));
        System.out.println(isVPS(stack) ? "YES" : "NO");
    }

    private static boolean isVPS(Stack<Character> stack) {
        int closedTime = 0;
        while (!stack.empty()) {
            if (stack.pop() == '(') {
                if (closedTime > 0)
                    closedTime--;
                else
                    return false;
            } else
                closedTime++;
        }
        return closedTime == 0;
    }

    public static void main(String[] args) {
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(System.in))) {
            int numberOfInput = Integer.parseInt(reader.readLine());
            for (int i = 0; i < numberOfInput; i++)
                command(reader.readLine());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: baekjoon/src/tags/data_structures/No9012.kt
package tags.data_structures

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.Stack

/**
 * https://www.acmicpc.net/problem/9012
 */
fun main() {
    val reader = BufferedReader(InputStreamReader(System.`in`))
    val writer = BufferedWriter(OutputStreamWriter(System.out))

    val numberOfInput = reader.readLine().toInt()
    repeat(numberOfInput) {
        val stack = Stack<Char>()
        var isValid = true

        val parentheses = reader.readLine()
        for (parenthesis in parentheses) {
            if (parenthesis == '(')
                stack.push(parenthesis)
            else {
                if (stack.isEmpty() || stack.pop() == ')') {
                    isValid = false
                    break
                }
            }
        }
        if (!stack.isEmpty())
            isValid = false

        writer.write(if (isValid) "YES" else "NO")
        writer.newLine()
    }
    writer.flush()
}
